File: implementation/Scala/src/main/scala/synapses/model/netElems/Activation.scala
package synapses.model.netElems

case class Activation(name: String,
                      f: Double => Double,
                      deriv: Double => Double,
                      inverse: Double => Double,
                      minMaxInVals: (Double, Double))

object Activation {

  def restrictedInput(activation: Activation, x: Double): Double =
    Math.max(
      Math.min(
        x,
        activation.minMaxInVals._2
      ),
      activation.minMaxInVals._1
    )

  def restrictedOutput(activation: Activation, y: Double): Double =
    Math.max(
      Math.min(
        y,
        activation.f(activation.minMaxInVals._2)
      ),
      activation.f(activation.minMaxInVals._1)
    )

  private def sigmoidF(x: Double): Double = {
    1.0 / (1.0 + Math.exp(-x))
  }

  val sigmoid: Activation =
    Activation(
      "sigmoid",
      sigmoidF,
      d => sigmoidF(d) * (1.0 - sigmoidF(d)),
      y => Math.log(y / (1.0 - y)),
      (-700.0, 20.0)
    )

  val identity: Activation =
    Activation(
      "identity",
      x => x,
      _ => 1.0,
      y => y,
      (-1.7976931348623157E308, 1.7976931348623157E308)
    )

  val tanh: Activation =
    Activation(
      "tanh",
      Math.tanh,
      d => 1.0 - Math.tanh(d) * Math.tanh(d),
      y => 0.5 * Math.log((1.0 + y) / (1.0 - y)),
      (-10.0, 10.0)
    )

  val leakyReLU: Activation =
    Activation(
      "leakyReLU",
      x => if (x < 0.0) {
        0.01 * x
      } else {
        x
      },
      d => if (d < 0.0) {
        0.01
      } else {
        1.0
      },
      y => if (y < 0.0) {
        y / 0.01
      } else {
        y
      },
      (-1.7976931348623157E308, 1.7976931348623157E308)
    )

  type ActivationSerialized = String

  def serialized(activation: Activation)
  : ActivationSerialized = activation.name

  def deserialized(activationSerialized: ActivationSerialized)
  : Activation =
    activationSerialized match {
      case "sigmoid" => Activation.sigmoid
      case "identity" => Activation.identity
      case "tanh" => Activation.tanh
      case "leakyReLU" => Activation.leakyReLU
    }

}

File: implementation/DotNet/Synapses/Model/NetElems/Activation.fs
module Synapses.Model.NetElems.Activation

open System

type Function = { name: string
                  f: float -> float
                  deriv: float -> float
                  inverse: float -> float
                  minMaxInVals: float * float }

type SerializableFunction = string

exception UnknownActivationF of string

let restrictedInput
        (activation: Function)
        (x: float)
        : float =
    let (minInVal, maxInVal) = activation.minMaxInVals
    max
      (min x maxInVal)
      minInVal

let restrictedOutput
        (activation: Function)
        (y: float)
        : float =
    let (minInVal, maxInVal) = activation.minMaxInVals
    max
      (min y (activation.f maxInVal))
      (activation.f minInVal)

let sigmoid_f (x: float): float =
    1.0 / (1.0 + Math.Exp(-x))

let sigmoid = { name = "sigmoid"
                f = sigmoid_f
                deriv = fun d ->
                    sigmoid_f d * (1.0 - sigmoid_f d)
                inverse = fun y ->
                    Math.Log(y / (1.0 - y))
                minMaxInVals = (-700.0, 20.0) }

let identity = { name = "identity"
                 f = fun x -> x
                 deriv = fun _ -> 1.0
                 inverse = fun y -> y
                 minMaxInVals = (-1.7976931348623157E308, 1.7976931348623157E308) }

let tanh = { name = "tanh"
             f = Math.Tanh
             deriv = fun d ->
                 1.0 - (Math.Tanh d) * (Math.Tanh d)
             inverse = fun y ->
                 0.5 * Math.Log((1.0 + y) / (1.0 - y))
             minMaxInVals = (-10.0, 10.0) }

let leakyReLU = { name = "leakyReLU"
                  f = fun x ->
                      match x with
                      | _ when x < 0.0 -> 0.01 * x
                      | _ -> x
                  deriv = fun d ->
                      match d with
                      | _ when d < 0.0 -> 0.01
                      | _ -> 1.0
                  inverse = fun y ->
                      match y with
                      | _ when y < 0.0 -> y / 0.01
                      | _ -> y
                  minMaxInVals = (-1.7976931348623157E308, 1.7976931348623157E308) }

let serialized
        (activF: Function):
        SerializableFunction =
    activF.name

let deserialized
    (s: SerializableFunction)
    : Function =
    match s with
    | "sigmoid" -> sigmoid
    | "identity" -> identity
    | "tanh" -> tanh
    | "leakyReLU" -> leakyReLU
    | _ -> raise (UnknownActivationF(s))
